### ISSUE TITLE
Cart recovery issues and silently consuming errors

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#30074]
## Describe the bug and add screenshots

Prestashop has a functionality that allows to send a prepared cart to customer via a special link. When he visits this link in front office, it will automatically assign proper customer, log him in if needed and load this cart. This link is used for example in **backoffice order** or custom **abandoned cart reminder** modules.

**This link looks something like this:**
`https://www.domain.com/cart?action=show&recover_cart=9&token_cart=4b2256bcb0b782757730be5d4e6b5e6e`

On every page load, Prestashop will check if `recover_cart` and `token_cart` are present in the URL try to load the cart. This happens in `recoverCart` function in `FrontController`.

However, there are some issues I found when testing this feature. If something fails, Prestashop won't notify the customer at all and it will look like if nothing happened, and customer wondering why his cart link is not working.

## What is wrong
- [x] After recovery, [you should be logged in](https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/orders/orders/add.md#send-an-email-to-the-customer-with-the-link-to-process-the-payment). Right now it assigns the customer to the context and you are able to continue checkout without password. Then it logs you in, but it should be right away - `needs testing.` Same issue reported here - https://github.com/PrestaShop/PrestaShop/issues/20977.

- [x] If the customer doesn't exist, it should be able to recover the cart. But if you send a cart to a customer already logged in, should it log him out? Or it should assing this cart to him? Or it should duplicate the cart and assign it to him? Simmilar issue - https://github.com/PrestaShop/PrestaShop/issues/10996

## What is fixed
- [x] If the cart or it's customer is deleted, the cart won't get loaded, but no error message is shown, and for example, just an empty cart page is shown. This can happen if customer opens an old link in his mailbox and you/module deleted the cart in the meantime. Fixed by adding an error in https://github.com/PrestaShop/PrestaShop/pull/30075.
- [x] If an order has been already made from this cart, the cart will be loaded at first, but discarded later and error message will appear in logs `Frontcontroller::init - Cart cannot be loaded or an order has already been placed using this cart`. Also, in this case, the current code destroys your current cart, because it loads the faulty one. Also, no error message here. Fixed by adding an error in https://github.com/PrestaShop/PrestaShop/pull/30075.

## Steps to reproduce

I made a test module to easily reproduce this - [tox_cartrecoverytest.zip](https://github.com/PrestaShop/PrestaShop/files/9824473/tox_cartrecoverytest.zip) This module can easily generate these cart recovery URLs, so you can test it quickly.

1. Install my test module. Go configure this module.
![test_tool](https://user-images.githubusercontent.com/6097524/196807176-fb9aa895-dbc5-4677-abea-71f078520403.jpg)
2. I prepared an easy generator for these URLs. You can enter any cart ID and it will generate the URL for you.
3. Then just copy it and visit it in FO.

## PrestaShop version(s) where the bug happened

develop

## PHP version(s) where the bug happened

8.0, 1.7.6.2

**Comment by HanaRebaiQA** [comment]
Hello @Hlavtox 

Thanks for your report.
It looks similar to this issue https://github.com/PrestaShop/PrestaShop/issues/10996 
What do you think?

**Comment by Hlavtox** [comment]
Hi @HanaRebaiQA, I am not sure, that issue is not very well explained and it's using not very good English. However, I think that the issue you mention will be also fixed by my PR.

In this issue, you have perfect explanation, deep information, test module. :-)

**Comment by HanaRebaiQA** [comment]
Hello 

Thanks @Hlavtox for your feedback!
I tried to reproduce your issue on develop using the your module [tox_cartrecoverytest (1).zip](https://github.com/PrestaShop/PrestaShop/files/9828344/tox_cartrecoverytest.1.zip) in order to generate the different URLs for the different cart ID generation. 
I have always this error.

![image](https://user-images.githubusercontent.com/92912932/196910260-b5687198-37fe-4c9c-b5fa-eb94908275b9.png)

https://user-images.githubusercontent.com/92912932/196910206-147d4e92-d423-4b3c-9a04-a5778f1a3aad.mp4

Is this the expected error ? 

Hello @drpark 
This issue seems to be similar to yours https://github.com/PrestaShop/PrestaShop/issues/10996 
Could you confirm it please ? In this case, can we close your issue and keep this one since it is more explanatory ?

Thanks all!

**Comment by Hlavtox** [comment]
@HanaRebaiQA You are reproducing it completely wrong. It's not about **CART RULES** but **CARTS**. `Prestashop has a functionality that allows to send a prepared cart to customer via a special link. This link is used for example in backoffice order or custom abandoned cart reminder modules.`

I even wrote in my test module, that you can find cart IDs on "Shopping cart" page.

**I provided exact steps to reproduce in this issue:**
![Snímek obrazovky 2022-10-20 115229](https://user-images.githubusercontent.com/6097524/196917182-8fcb15ca-2011-4c80-9be2-4d793c8f944b.jpg)

**Comment by HanaRebaiQA** [comment]
Hello @Hlavtox 

Yes! you are right! But i also tried to reproduce with Carts in `Shopping carts` page on develop and php 8.0 and i was unable to reproduce it. I have always the same message.

Did I miss something?


https://user-images.githubusercontent.com/92912932/196958906-8ed8828f-f6a5-4e80-9958-0de954afb61b.mp4

Thank You!

**Comment by Hlavtox** [comment]
@HanaRebaiQA Yes, that is the error, that it should show an error message instead of an empty cart. Also, you cannot just change the id in the URL, the token is unique. That's why I provided the test module to easily generate the URLs.

Let's try a different approach.

### Scenario 1
0. Clear cookies.
1. Go to FO, add something to cart, finish step 1 in checkout. See that in BO, a cart was created in "Shopping carts" page.
2. Go to my module, generate link for this cart. Imagine that you send this to your customer. 
3. Open it in another browser, your cart will get loaded. This works. ✅

### Scenario 2
0. Clear cookies.
1. Go to FO, add something to cart, finish step 1 in checkout. See that in BO, a cart was created in "Shopping carts" page.
2. Go to my module, generate link for this cart. Imagine that you send this to your customer.
4. Delete the cart from "Shopping carts".
5. Open the link - no error message, only empty cart displayed - that's wrong, you should be notified that this cart is not usable anymore. 🔴

### Scenario 3
0. Clear cookies.
1. Go to FO, add something to cart, finish step 1 in checkout. See that in BO, a cart was created in "Shopping carts" page.
2. Go to my module, generate link for this cart. Imagine that you send this to your customer.
3. Go to "Customers" in backoffice and delete the customer related to this cart.
4. Open the link - no error message, only empty cart displayed - that's wrong, you should be notified that this cart is not usable anymore. 🔴

### Scenario 4
0. Clear cookies.
1. Go to FO, add something to cart, finish step 1 in checkout. See that in BO, a cart was created in "Shopping carts" page.
2. Go to my module, generate link for this cart. Imagine that you send this to your customer.
3. Finish the order in FO.
4. Open the link - no error message, only empty cart displayed - that's wrong, you should be notified that this cart is not usable anymore, because it was used in an order already. 🔴

**Comment by NeOMakinG** [comment]
I reproduced the behavior, but we need to verify with the @PrestaShop/product-team if the error message is the behavior we want

@MatShir would you have some time to discuss it?

**Comment by Hlavtox** [comment]
**Another things found:**
- After recovery, you should be probably logged in. Right now it assigns the customer to the context and you are able to continue checkout without password. Then it logs you in, but it should be right away.
- If the customer doesn't exist, it should be able to recover the cart. But if you send a cart to a customer already logged in, should it log him out? Or it should assing this cart to him? Or it should duplicate the cart and assign it to him?

**Things reproduced, clear and fixed by PR:**
- Cart already ordered - show error. Do not change anything in the context.
- Cart non existent - show error. Do not change anything in the context.

**Comment by Hlavtox** [comment]
Fixed!
